### ISSUE TITLE
Running Malcolm in ECS Fargate

### ISSUE THREAD

**Issue by ryfestag** [idaholab/Malcolm#223]
We are interested in seeing if we can get Malcom working in ECS Fargate. There are a few constraints we need to work around in this context, and are hoping for some guidance/input on whether they are feasible:

1. Very few [kernel capabilities](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_KernelCapabilities.html) are supported.  In particular Fargate only allows SYS_PTRACE. Various containers currently also add IPC_LOCK, NET_ADMIN, NET_RAW, SYS_ADMIN, SYS_RESOURCE, In our case, we're primarily planning to simply upload files (vs. live capture). What are the implications on running Malcolm if these capabilities are not actually supported (does it simply degrade functionality? Or will it be non-functional?)
2. ECS doesn't support individual file bind mounts like are used. This complicates the process of mounting configuration files into the containers (particularly those containing secrets). I saw some work had been done to enable some secrets via env variables, but I'm having trouble finding full documentation. Ideally we'd like to use AWS Secrets Manager to securely expose the parameters as environment variables into the containers. 

Any help understanding the ramifications of Item 1 and the ability to do Item 2 would be greatly appreciated!

**Comment by mmguero** [comment]
1. Most of the issues you'd have here are going to be related to live capture, so if you're not doing that then you probably won't have any issues. The only one I'm not sure of is that some of documentation I read about OpenSearch (and ElasticSearch from which it was forked) talk about having that capability turned on for that container, "for increasing memlock limit on the node". You could try it without and see what happens. If it doesn't work you could look at hosting an [external opensearch instance](https://idaholab.github.io/Malcolm/docs/opensearch-instances.html#OpenSearchInstance) and pointing malcolm at that.
2. The documentation on the environment variable files [can be found here](https://idaholab.github.io/Malcolm/docs/malcolm-config.html#MalcolmConfigEnvVars). These values are populated in the files for you when you run the [Malcolm configuration](https://idaholab.github.io/Malcolm/docs/malcolm-hedgehog-e2e-iso-install.html#MalcolmConfig) and answer the questions there. If you'd like to expose those environment variables using some other mechanism, you're welcome to do that, as long as they're available for the containers when they run I don't see why that wouldn't work. However, I do see you running into some major issues with not being able to bind-mount files. There are a lot of files which have to be provided to Malcolm, some of which are created at runtime (and as such exist in directories mounted as volumes), others are created beforehand during the [auth setup](https://idaholab.github.io/Malcolm/docs/authsetup.html#AuthSetup) phase and/or during configuration and startup. Maybe some of the documentation on deploying Malcolm using [Kubernetes](https://idaholab.github.io/Malcolm/docs/kubernetes.html#PVC) (or [EKS, specifically](https://idaholab.github.io/Malcolm/docs/kubernetes-eks.html#KubernetesEKS)) dealing with the persistentvolumeclaims could at least show you the areas that files would need to exist from. With K8s, at least, directories are done using PeristentVolumes and PersistentVolumeClaims, and the other configuration files are done with [ConfigMaps](https://github.com/idaholab/Malcolm/blob/b5e9100f2da6d9f3c2b7db6a7db30fcf9423379b/scripts/malcolm_kubernetes.py#L39-L160). I can't speak to how Fargate ECS works. In my bleeding-edge development fork which has not yet been released and merged back into this repository, I do have a [packer template](https://mmguero-dev.github.io/Malcolm/docs/third-party-envs.html#AWSAMI) for building an ECS AMI, but I don't think that's the same thing.

TL;DR:

1. If you're not doing live capture, it will *probably* work, give it a shot.
2. [environment variable files](https://idaholab.github.io/Malcolm/docs/malcolm-config.html#MalcolmConfigEnvVars) are set during [Malcolm configuration](https://idaholab.github.io/Malcolm/docs/malcolm-hedgehog-e2e-iso-install.html#MalcolmConfig). You may be able to mimic what we're doing with K8s.

Fargate isn't on our roadmap at the moment, but we'd welcome any updates or things you come up with as comments on this issue.

**Comment by ryfestag** [comment]
Great, thanks for the quick reply! I thought that would be the case regarding the capabilities, given we're just doing the file upload use case for now, but wanted to make sure there wasn't anything obvious. We plan to use the AWS Managed OpenSearch as an external instance service, so shouldn't need to run those containers.

Regarding the documentation on env files: I guess I expected there to be an env for things like the OpenSearch user and password. Based on the link, that isn't the case (it is coming from that curlrc file that gets mounted in). I'll have to take a closer look at how the ConfigMaps work and whether there is some analagous process we can use. The only other alternative I can think of is to write  a script that goes through the docker-compose, finds the target directories that files use, and copy them out. Then drop the actual target file in, and expose that whole structure through EFS Access Points (we can mount directories from EFS into the containers similarly to PVCs). 

Yea, the packer work (while definitely interesting) isn't going to be applicable here. I'll take a closer look at the ConfigMaps and PVCs you are doing in Kubernets, and seeing whether we can do something similar in ECS. I suspect that the ConfigMaps are getting around the core issue of putting individual files into the containers (but I'm not K8s expert).

**Comment by mmguero** [comment]
What you're talking about doing with populating those files that way is not dissimilar to what we're doing for Kubernetes support. The [Kubernetes manifests](https://github.com/idaholab/Malcolm/tree/main/kubernetes) will show what services use what volumes, along with that like I posted about the [ConfigMaps](https://github.com/idaholab/Malcolm/blob/b5e9100f2da6d9f3c2b7db6a7db30fcf9423379b/scripts/malcolm_kubernetes.py#L39-L160). There is also some other special startup code in all of our containers that is hooked as the entrypoint: any directories that are named with either of the values specified by [CONFIG_MAP_DIR here](https://github.com/idaholab/Malcolm/blob/b5e9100f2da6d9f3c2b7db6a7db30fcf9423379b/config/process.env.example#L6-L7) will [automatically be rsynced to the parent directory](https://github.com/idaholab/Malcolm/blob/b5e9100f2da6d9f3c2b7db6a7db30fcf9423379b/shared/bin/docker-uid-gid-setup.sh#L24-L70) in the container's startup. That may or may not be helpful to you. Good luck.
